core/storage/hamt/hamt.cpp

```diff
@@ -0,0 +1,247 @@
+/**
+ * Copyright Soramitsu Co., Ltd. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include "storage/hamt/hamt.hpp"
+
+#include "codec/cbor/cbor.hpp"
+#include "common/which.hpp"
+#include "crypto/blake2/blake2b160.hpp"
+#include "crypto/murmur/murmur.hpp"
+
+OUTCOME_CPP_DEFINE_CATEGORY(fc::storage::hamt, HamtError, e) {
+  using fc::storage::hamt::HamtError;
+  switch (e) {
+    case HamtError::EXPECTED_CID:
+      return "Expected CID";
+    case HamtError::NOT_FOUND:
+      return "Not found";
+    case HamtError::MAX_DEPTH:
+      return "Max depth exceeded";
+    default:
+      return "Unknown error";
+  }
+}
+
+namespace fc::storage::hamt {
+  using fc::common::which;
+
+  const CID kDummyCid({}, {}, libp2p::multi::Multihash::create({}, {}).value());
+
+  // assuming 8-bit indices
+  auto keyToIndices(const std::string &key, int n = -1) {
+    std::vector<uint8_t> key_bytes;
+    for (auto c : key) {
+      key_bytes.push_back(c);
+    }
+    auto hash = crypto::murmur::hash(key_bytes);
+    std::vector<size_t> result;
+    for (auto i = n == -1 ? hash.begin() : hash.end() - n + 1; i != hash.end();
+         ++i) {
+      result.push_back(*i);
+    }
+    return result;
+  }
+
+  auto bitToIndex(UBigInt bits, size_t max) {
+    size_t index = 0;
+    for (size_t i = 0; i < max; i++) {
+      if (bits & 1) {
+        ++index;
+      }
+      bits >>= 1;
+    }
+    return index;
+  }
+
+  auto consumeIndex(gsl::span<const size_t> indices) {
+    return indices.subspan(1);
+  }
+
+  outcome::result<CID> Node::cid(gsl::span<const uint8_t> encoded) {
+    OUTCOME_TRY(hash_raw, crypto::blake2b::blake2b_256(encoded));
+    OUTCOME_TRY(hash,
+                libp2p::multi::Multihash::create(
+                    libp2p::multi::HashType::blake2b_256, hash_raw));
+    return CID(CID::Version::V1, libp2p::multi::MulticodecType::DAG_CBOR, hash);
+  }
+
+  outcome::result<CID> Node::cid() const {
+    OUTCOME_TRY(encoded, codec::cbor::encode(*this));
+    return cid(encoded);
+  }
+
+  Hamt::Hamt(std::shared_ptr<ipfs::IpfsDatastore> store, Node::Ptr root)
+      : store_(std::move(store)), root_(root) {}
+
+  Hamt::Hamt(std::shared_ptr<ipfs::IpfsDatastore> store, const CID &root)
+      : store_(std::move(store)), root_(root) {}
+
+  outcome::result<void> Hamt::set(const std::string &key,
+                                  gsl::span<const uint8_t> value) {
+    OUTCOME_TRY(loadItem(root_));
+    return set(*boost::get<Node::Ptr>(root_), keyToIndices(key), key, value);
```

Comment:
Should here not be a check whether root_ contains Node::Ptr?

Reply:
`Hamt::loadItem(Node::Item)` ensures that node is either `Node::Ptr` or `Node::Leaf`.
`root_` is never `Node::Leaf`.